src/parser.rs

```diff
@@ -917,7 +917,26 @@ impl<'run, 'src> Parser<'run, 'src> {
             first: *line,
           }));
         }
-        attributes.insert(attribute, name.line);
+        match attribute {
```

Comment:
How about something like this:

```rust
let arguments = if self.accepted(ParenL)? {
  let mut arguments = Vec::new();
  while self.accepted(StringToken)? {
    arguments.push(self.parse_string_literal()?);
  }
  self.expect(ParenR)?;
  Some(arguments)
} else {
  None
};

let attribute = if let Some(arguments) = arguments {
  attribute.with_arguments(arguments)?
} else {
  attribute
};
```

`Attribute::with_arguments` then returns an error if the attribute doesn't take the given number of arguments.

There are some places in the code which take the approach of parsing different depending on a keyword, but this leads to confusing error messages.

Reply:
I revised the parsing but one test fails.

```shell
failures:

---- attributes::multiple_attributes_one_line_error_message stdout ----
Bad stderr: Diff < left / right > :
<"error: Expected ']', ',', or '(', but found identifier\n ——▶ justfile:1:17\n  │\n1 │ [macos, windows linux]\n  │                 ^^^^^\n"
>"error: Expected ']' or ',', but found identifier\n ——▶ justfile:1:17\n  │\n1 │ [macos, windows linux]\n  │                 ^^^^^\n"
```
It fails because recipe attributes generally accept arguments now. So maybe the error message should just be revised for the new kind? And then if you write e.g. `[linux("arg")]` it should say something like "expected 0 arguments got 1"?

Not sure what the preferred way of fixing it is.

Reply:
These tests can just be updated. When the parser encounters an unexpected token, in prints out the tokens that would have been acceptable. Since now, `(` is accepted after an attribute name, the `(` is part of the acceptable tokens.